src/ether.c

```diff
@@ -98,6 +108,603 @@ int ETHEREventCount = 0;
+ **********************************************************************/
+LispPTR ether_reset(LispPTR args[])
+{
+  if (nethubHost == NULL) {
+    return (NIL);
+  }
+  log_debug(("ether_reset() - begin\n"));
+  
+  if (ether_fd < 0) {
+    log_debug(("ether_reset() - end -> NIL\n\n"));
+    log_info(("ether_reset() -> NIL\n"));
+    return (NIL);
+  }
+  ether_bsize = 0; /* deactivate receiver */
+  log_debug(("ether_reset() - end -> ATOM_T\n\n"));
+  log_info(("ether_reset() -> ATOM_T\n"));
+  return (ATOM_T);
+}
+
+
+/************************************************************************/
+/*									*/
+/*			  e t h e r _ g e t (175/74/2)			*/
+/*									*/
+/*	Set up the Ethernet driver to receive a packet.  The driver	*/
+/*	first tries to read any pending packet from the net, and if	*/
+/*	there is one, ether_get returns T.  If there is no pending	*/
+/*	packet, the failing read sets us up to get an interrupt when	*/
+/*	a packet DOES arrive, and ether_get returns NIL.		*/
+/*									*/
+/*		args[0] Length of the buffer we're passed		*/
+/*		args[1] LISP address of a packet buffer			*/
+/*									*/
+/*	sets ether_buf to the buffer address, for check_ether's use	*/
+/*	sets ether_bsize to the buffer size.  ether_bsize>0 means	*/
+/*	it's OK to read packets from the network on interrupt.		*/
+/*									*/
+/************************************************************************/
+LispPTR ether_get(LispPTR args[])
+{
+  if (nethubHost == NULL) {
+    return (NIL);
+  }
+  log_debug(("ether_get() - begin\n"));
+
+  u_char *target = (u_char *)NativeAligned2FromLAddr(args[1]);
+  int maxByteCount = 2 * (0xFFFF & args[0]); /* words to bytes */
+  log_debug(("  target = 0x%016lX maxBytecount: %d bytes\n", (unsigned long)target, maxByteCount));
+
+  sigset_t signals;
+  sigemptyset(&signals);
+  sigaddset(&signals, SIGIO);
+  sigprocmask(SIG_BLOCK, &signals, NULL);
+
+  ether_buf = target;
+  ether_bsize = maxByteCount;
+
+  int receivedBytes = recvPacket();
+
+  sigprocmask(SIG_UNBLOCK, &signals, NULL);
+
+  log_debug(("ether_get() ::  receivedBytes: %d bytes\n", receivedBytes));
+  if (receivedBytes <= 0) {
+    log_debug(("ether_get() - end -> NIL\n\n"));
+    return (NIL);
+  }
+
+  log_debug(("ether_get() - end -> ATOM_T\n\n"));
+  log_info(("ether_get(): received packet, len = %d bytes\n", receivedBytes));
+  return (ATOM_T);
+}
+
+
+
+/**********************************************************************
+ *	ether_send(args) 175/75/2 max_words,buffer_addr
+ *	send a packet
+ **********************************************************************/
+LispPTR ether_send(LispPTR args[])
+{
+  if (nethubHost == NULL) {
+    return (NIL);
+  }
+  log_debug(("ether_send() - begin\n"));
+
+  u_char *source = (u_char *)NativeAligned2FromLAddr(args[1]);
+  int byteCount = 2 * (0xFFFF & args[0]); /* words to bytes */
+
+  log_debug(("   source = 0x%08X , bytecount: %d bytes\n", (unsigned int)source, byteCount));
+
+#ifdef BYTESWAP
```

Comment:
I need to check this, but it feels wrong.  The Lisp code should have assembled the packet in network byte order.  If it turns out that swapping is necessary, I'd suggest copy to a temporary buffer, swap, send, and discard rather than re-swapping (or swap on the fly during the copy)

Reply:
left it this way, swapping bytes is absolutely necessary (it took me a long time to understand how this is swapped, as the data pointer points to the 2nd 16-bit word of a 32-bit doubleword, so the first 2 packet bytes are in fact _before_ the pointer and the first 2 bytes pointed to are not part of the packet)

Reply:
for copying the packet to a buffer and then swapping the buffer:

- that's what i did initially, the i saw that the _swap-use-swap_-pattern was used in the modules `inet.c`, `dsk.c`, `chardev.c` or `unixcomm.c`, so i switched to this pattern...
- and so i left it this way, also as the data is partly _before_ the original data pointer

Reply:
The pointer is subject to interpretation... it's not intended to be a pointer that can be used without manipulation by the `GETxxxx` macros to point it at the appropriate place for the endianness of the machine it's running on.
I'm OK merging it as is, but this is an area that we should look at very carefully once that's done.